File: archive/pulumi-ts/Pr7298/lib/xray-stack.ts
/**
 * xray-stack.ts
 *
 * Defines AWS X-Ray distributed tracing configuration.
 *
 * Features:
 * - Sampling rules (10% of requests)
 * - X-Ray groups for filtering traces by environment
 * - Service map for visualizing microservices architecture
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface XrayStackArgs {
  environmentSuffix: string;
  sampleRate?: number; // Default 0.1 (10%)
  tags?: pulumi.Input<{ [key: string]: string }>;
}

export class XrayStack extends pulumi.ComponentResource {
  public readonly samplingRule: aws.xray.SamplingRule;
  public readonly xrayGroup: aws.xray.Group;

  constructor(
    name: string,
    args: XrayStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:xray:XrayStack', name, args, opts);

    const { environmentSuffix, sampleRate = 0.1, tags } = args;

    // Create X-Ray sampling rule (10% by default)
    this.samplingRule = new aws.xray.SamplingRule(
      `cicd-sampling-rule-${environmentSuffix}`,
      {
        ruleName: `cicd-pipeline-${environmentSuffix}`,
        priority: 1000,
        version: 1,
        reservoirSize: 1, // Number of requests to trace per second before applying fixed rate
        fixedRate: sampleRate, // 0.1 = 10% of requests
        urlPath: '*', // Trace all paths
        host: '*', // Trace all hosts
        httpMethod: '*', // Trace all HTTP methods
        serviceName: '*', // Trace all services
        serviceType: '*', // Trace all service types
        resourceArn: '*', // Trace all resources
        attributes: {
          environment: environmentSuffix,
        },
        tags: {
          ...tags,
          Name: `cicd-sampling-rule-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create X-Ray group for filtering traces
    this.xrayGroup = new aws.xray.Group(
      `cicd-xray-group-${environmentSuffix}`,
      {
        groupName: `cicd-pipeline-${environmentSuffix}`,
        filterExpression: `annotation.environment = "${environmentSuffix}"`,
        insightsConfiguration: {
          insightsEnabled: true, // Enable X-Ray Insights
          notificationsEnabled: false, // Disable notifications
        },
        tags: {
          ...tags,
          Name: `cicd-xray-group-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    this.registerOutputs({
      samplingRuleName: this.samplingRule.ruleName,
      samplingRuleArn: this.samplingRule.arn,
      xrayGroupName: this.xrayGroup.groupName,
      xrayGroupArn: this.xrayGroup.arn,
    });
  }
}

File: archive/cdk-py/Pr8009/lib/xray_stack.py
"""X-Ray tracing configuration stack"""

from typing import Optional
import aws_cdk as cdk
from aws_cdk import (
    aws_xray as xray,
)
from constructs import Construct


class XRayStackProps(cdk.StackProps):
    """Properties for XRayStack"""

    def __init__(self, environment_suffix: Optional[str] = None, **kwargs):
        super().__init__(**kwargs)
        self.environment_suffix = environment_suffix


class XRayStack(Construct):
    """Stack for X-Ray tracing configuration"""

    def __init__(self, scope: Construct, construct_id: str, props: Optional[XRayStackProps] = None, **kwargs):
        super().__init__(scope, construct_id)

        env_suffix = props.environment_suffix if props else 'dev'

        # Create X-Ray sampling rule with 0.1 sampling rate (10%)
        sampling_rule = xray.CfnSamplingRule(
            self,
            f"PaymentSamplingRule-{env_suffix}",
            sampling_rule=xray.CfnSamplingRule.SamplingRuleProperty(
                fixed_rate=0.1,
                host="*",
                http_method="*",
                priority=100,
                reservoir_size=1,
                resource_arn="*",
                rule_name=f"pay-sample-{env_suffix}",
                service_name="*",
                service_type="*",
                url_path="*",
                version=1
            )
        )

        # Create X-Ray group for payment processing
        payment_group = xray.CfnGroup(
            self,
            f"PaymentGroup-{env_suffix}",
            group_name=f"payment-processing-{env_suffix}",
            filter_expression='service("payment-api") OR service("payment-lambda")'
        )

        # Create X-Ray group for API Gateway
        api_group = xray.CfnGroup(
            self,
            f"ApiGatewayGroup-{env_suffix}",
            group_name=f"api-gateway-{env_suffix}",
            filter_expression='service("AWS::ApiGateway::Stage")'
        )

        # Create X-Ray group for Lambda functions
        lambda_group = xray.CfnGroup(
            self,
            f"LambdaGroup-{env_suffix}",
            group_name=f"lambda-functions-{env_suffix}",
            filter_expression='service("AWS::Lambda") OR service("AWS::Lambda::Function")'
        )
